data gen source

sophiecuiy opened the pull request:
## What
<!--
* Describe what the change is solving. Link all GitHub issues related to this change.
-->
creating new datagen source for speed testing.

sophiecuiy committed:
progress checkpoint

sophiecuiy committed:
file, config clean up

sophiecuiy committed:
check and discover

sophiecuiy committed:
partitioning for read

sophiecuiy committed:
partitionReader progress

sophiecuiy committed:
read done, jsonb/int type mismatch

sophiecuiy committed:
read working, configspec update

sophiecuiy committed:
partition logic fix, add logging, add shutdown

sophiecuiy committed:
not stopping, changed condition to max rows from run duration

sophiecuiy committed:
read completed for 1 partition

sophiecuiy committed:
multi-partition implementation STDIO

sophiecuiy commented:
/format-fix

octavia-squidington-iii committed:
chore: auto-fix lint and format issues

sophiecuiy committed:
format fix

# Conflicts:
#	airbyte-integrations/connectors/source-datagen/src/main/kotlin/io/airbyte/integrations/source/datagen/flavor/increment/IncrementDataGenerator.kt

sophiecuiy committed:
format fix

sophiecuiy commented:
/format-fix

octavia-squidington-iii committed:
chore: auto-fix lint and format issues

sophiecuiy committed:
undo dev-null change

sophiecuiy committed:
documentation start

sophiecuiy committed:
added changelog

sophiecuiy committed:
max record logic fix

sophiecuiy committed:
starting schema config implementation

sophiecuiy committed:
Revert "starting schema config implementation"

This reverts commit 748d4fa1d6f45df45273654a3c96aeae06774ba9.

sophiecuiy committed:
single stream SPEED

sophiecuiy committed:
cdk version decl change

sophiecuiy committed:
cdk version decl change

sophiecuiy committed:
cdk version decl change 2

sophiecuiy committed:
kjetland dependency

sophiecuiy committed:
adding fields to schema

mwbayley reviewed (changes requested):
I encountered a bunch of MySQL copypasta. Remove this and I'll re-review!

sophiecuiy committed:
mysql stuff cleanup

sophiecuiy committed:
acceptance test config

octavia-squidington-iii committed:
chore: auto-fix lint and format issues

sophiecuiy committed:
cdk version push

# Conflicts:
#	airbyte-cdk/bulk/version.properties
#	airbyte-integrations/connectors/source-datagen/src/main/kotlin/io/airbyte/integrations/source/datagen/flavor/increment/IncrementFlavor.kt

sophiecuiy committed:
Bump cdk version to 0.1.40

sophiecuiy committed:
Merge branch 'master' into sophie/source_no_op

sophiecuiy committed:
Update changelog for version 0.1.45

Added additional AirbyteSchemaTypes in version 0.1.45.

sophiecuiy committed:
Delete acceptance-test-config.yml

mwbayley commented on airbyte-integrations/connectors/source-datagen/src/main/kotlin/io/airbyte/integrations/source/datagen/partitionops/DataGenPartitionReader.kt:
```diff
@@ -0,0 +1,118 @@
+
+        val configuration = sharedState.configuration
+        val sourceDataGenerator = configuration.flavor.dataGenerator
+        val recordCountPerPartition = configuration.maxRecords / partition.modulo
```
When `maxRecords` doesn't divide evenly by `partition.modulo`, we'll end up a few records short. It's not a huge deal from a correctness or performance testing perspective, but it doesn't inspire confidence in users. Can you think of an easy way to always generate the exact number of total records that were requested? This solution can be delivered in a follow-up.

mwbayley commented on airbyte-integrations/connectors/source-datagen/src/main/kotlin/io/airbyte/integrations/source/datagen/partitionops/DataGenPartitionReader.kt:
```diff
@@ -0,0 +1,118 @@
+            if (runComplete.get()) {
+                DataGenStreamState.completeState
+            } else {
+                com.fasterxml.jackson.databind.node.JsonNodeFactory.instance.objectNode()
```
you could add a comment saying that we don't support restoring from state, so the checkpointed state is only used to indicate completion, and other values are meaningless/unused

mwbayley commented on airbyte-integrations/connectors/source-datagen/src/main/resources/application.yml:
```diff
@@ -0,0 +1,145 @@
+    output:
+      buffer-byte-size-threshold-for-flush: 8192
+    extract:
+      jdbc:
+        mode: concurrent
+        with-sampling: true
+        table-sample-size: 1024
+        throughput-bytes-per-second: 10000000
+        min-fetch-size: 10
+        default-fetch-size: 1024
+        max-fetch-size: 1000000000
+        memory-capacity-ratio: 0.6
+        estimated-record-overhead-bytes: 16
+        estimated-field-overhead-bytes: 16
+        namespace-kind: CATALOG_AND_SCHEMA
+    check:
+      jdbc:
+        queries:
+          - >-
+            SELECT 1 FROM dual WHERE 1 = 0;
```
Most of this should be removed. Certainly the query should be. Provide an empty list there if it's required. Do similarly for the other unused configuration.

mwbayley commented on airbyte-integrations/connectors/source-datagen/src/main/resources/application.yml:
```diff
@@ -0,0 +1,145 @@
+
+    exception-classifiers:
+      regex:
+        # The following rules are for the RegexExceptionClassifier [0] which are applied
+        # sequentially on a Throwable's message [1] and its nested messages by cause [2].
+        #
+        # This classifier's rules are applied ahead of the JdbcExceptionClassifier's further down.
+        #
+        # [0] https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/bulk/core/base/src/main/kotlin/io/airbyte/cdk/output/ExceptionClassifier.kt
+        # [1] https://docs.oracle.com/javase/8/docs/api/java/lang/Throwable.html#getMessage--
+        # [2] https://docs.oracle.com/javase/8/docs/api/java/lang/Throwable.html#getCause--
```
remove Oracle references

mwbayley commented on airbyte-integrations/connectors/source-datagen/src/main/resources/application.yml:
```diff
@@ -0,0 +1,145 @@
+
+          - pattern: (?i).*connection is not available, request timed out after.*
+            input-example: >-
+              java.sql.SQLTransientConnectionException: HikariPool-x - 
+              Connection is not available, request timed out after 10 ms
+            error: transient
+            group: Hikari Connection Pool Timeout
+            output: The sync encountered a database read failure due to a connection timeout, will retry.
+            reference-links: https://docs.oracle.com/javase/9/docs/api/java/sql/SQLTransientConnectionException.html
+
+          - pattern: (?i).*is unrecognized or represents more than one time zone.*
+            input-example: >-
+              java.sql.SQLException: The server time zone value 'PDT' is unrecognized or represents more than one time zone. 
+              You must configure either the server or JDBC driver (via the serverTimezone configuration property) to use a more specifc time zone value 
+              if you want to utilize time zone support.
+            error: config
+            group: MySQL Timezone Configuration Issue
+            output: >-
+              The sync encountered a timezone configuration issue. Please configure your database with the correct timezone as specified in the error message.
+              For more details, refer to: https://dev.mysql.com/doc/refman/8.4/en/time-zone-support.html.
+            reference-links: https://docs.oracle.com/javase/8/docs/api/index.html?java/io/SQLException.html
+
+          - pattern: (?s).*vttablet.*?rpc\s+error.*?code\s*=\s*Unavailable.*?desc\s*=\s*error\s+reading\s+from\s+server\s*:\s*EOF.*
+            input-example: >-
+              java.lang.RuntimeException: java.lang.RuntimeException: java.sql.SQLException: target: ai-solutions.-.primary: 
+              vttablet: rpc error: code = Unavailable desc = error reading from server: EOF
+            error: transient
+            group: MySQL Vitess Connection Error
+            output: The sync encountered a connection issue while reading from the MySQL Vitess database, will retry.
+            reference-links: https://vitess.io/docs/archive/13.0/user-guides/configuration-basic/troubleshooting.html
+
+          - pattern: (?i).*exception occurred in the change event produce.*
+            input-example: >-
+              java.lang.RuntimeException: org.apache.kafka.connect.errors.ConnectException: 
+              An exception occurred in the change event producer. This connector will be stopped.
+            error: config
+            group: MySQL Connector Error
+            output: >-
+              The sync encountered an unexpected error in the change event producer and has stopped. Please check the logs for details and troubleshoot accordingly.
+            reference-links: https://docs.oracle.com/javase/9/docs/api/java/lang/RuntimeException.html
```
remove - should be empty list

mwbayley commented on airbyte-integrations/connectors/source-datagen/src/main/resources/application.yml:
```diff
@@ -0,0 +1,145 @@
+        #
+        rules:
+          ## JDBC RULE TEMPLATE
+          # code: Required, MySQL vendor error code.
```
not MySQL

mwbayley commented on airbyte-integrations/connectors/source-datagen/src/test/resources/expected-spec.json:
this too - it's gotta go!

mwbayley reviewed (commented):
Structure is there but lots of cleanup needed.

sophiecuiy committed:
single stream clean up

sophiecuiy committed:
more cleanup

sophiecuiy committed:
flavor streamlining

sophiecuiy committed:
format fix

sophiecuiy committed:
format fix

octavia-squidington-iii committed:
chore: auto-fix lint and format issues

mwbayley commented on airbyte-integrations/connectors/source-datagen/src/main/kotlin/io/airbyte/integrations/source/datagen/DataGenSourceConfigurationSpecification.kt:
```diff
@@ -38,9 +38,9 @@ class DataGenSourceConfigurationSpecification : ConfigurationSpecification() {
+    @JsonSchemaTitle("Max Concurrency")
     @JsonSchemaDescription(
-        "Maximum number of concurrent queries to the database. Leave empty to let Airbyte optimize performance."
+        "Maximum number of concurrent partitions running at once. Leave empty to let Airbyte optimize performance."
```
Partitions are an internal concept. "Maximum number of concurrent data generators"?

mwbayley commented on airbyte-integrations/connectors/source-datagen/src/main/kotlin/io/airbyte/integrations/source/datagen/DataGenSourceMetadataQuerier.kt:
```diff
@@ -47,8 +38,7 @@ class DataGenSourceMetadataQuerier() : MetadataQuerier {
     ): List<List<String>> {
-        val flavor = flavors[streamID.namespace]
-        return listOf(flavor?.primaryKey ?: emptyList())
+        return listOf(configuration.flavor.primaryKey)
```
See how we're not using the `streamId` here? Each flavor should have a primary key for each table. PrimaryKey could be a map. That would create a lot of spooky coupling between that and the schemas, though. Alternatively, we could create a data class to hold primary key and schema, then each flavor would have a map from table name to an object of that data class.

mwbayley commented on airbyte-integrations/connectors/source-datagen/src/main/kotlin/io/airbyte/integrations/source/datagen/flavor/increment/IncrementDataGenerator.kt:
```diff
@@ -7,39 +7,25 @@ package io.airbyte.integrations.source.datagen.flavor.increment
-                }
-            recordData[field.id] = value
-        }
+        recordData["id"] =
+            FieldValueEncoder(incrementedID, IntegerFieldType.jsonEncoder as LongCodec)
+
+        recordData["boolean"] =
+            FieldValueEncoder(incrementedID % 2 == 1L, BooleanFieldType.jsonEncoder as BooleanCodec)
+
+        recordData["string"] =
+            FieldValueEncoder("string$incrementedID", StringFieldType.jsonEncoder as TextCodec)
+
```
I thought the increment flavor had just an `id`?

mwbayley replied:
Making the rows wider might make the data more useful for performance testing. The boolean doesn't seem to add much, though.

mwbayley reviewed (commented):
Very close! Just a few more things.

sophiecuiy committed:
format and pk fix

sophiecuiy committed:
format fix

sophiecuiy committed:
format fix

sophiecuiy commented:
/format-fix

mwbayley commented on airbyte-integrations/connectors/source-datagen/src/main/kotlin/io/airbyte/integrations/source/datagen/flavor/increment/IncrementFlavor.kt:
```diff
@@ -24,11 +22,9 @@ data object IncrementFlavor : Flavor {
                 )
         )
-    override val primaryKey = listOf("id")
+    override val primaryKey = mapOf(incrementTableName to listOf(listOf(("id"))))
```
nit: now it's `primaryKeys`

mwbayley reviewed (approved):
🎉🎉🎉🎉🎉🎉 Nice work!

One little naming change left, but approving now so you can merge once that's in.

sophiecuiy committed:
naming fix

sophiecuiy committed:
Merge branch 'master' into sophie/source_no_op
